File: lang/c++/test/muse/algorithms/meta/SearchTest.cpp
#include <muse/algorithms/meta/Search.hpp>
#include <muse/util/SequenceBuilder.hpp>
#include <muse/util/TestRunner.hpp>

using namespace std;

bool testBinarySearch() {
    size_t size = 32768;

    vector<int> arr(size);
    SequenceBuilder::packIncreasing(arr);

    if (Search::binarySearch(arr, -1) != size) {
        return false;
    }

    if (Search::binarySearch(arr, 2'147'483'647) != size) {
        return false;
    }

    for (size_t i = 0; i < size; i++) {
        if (Search::binarySearch(arr, arr[i]) != i) {
            return false;
        }
    }

    return true;
}

bool testLinearSearch() {
    size_t size = 32768;

    vector<int> arr(size);
    SequenceBuilder::packIncreasing(arr);

    if (Search::linearSearch(arr, -1) != size) {
        return false;
    }

    if (Search::linearSearch(arr, 2'147'483'647) != size) {
        return false;
    }

    for (size_t i = 0; i < size; i++) {
        if (Search::linearSearch(arr, arr[i]) != i) {
            return false;
        }
    }

    return true;
}

int main() {
    TestRunner::parseTest(testBinarySearch());

    TestRunner::parseTest(testLinearSearch());
}

File: lang/go/test/muse/algorithms/meta/Search/SearchTest.go
package main

import (
    "muse/algorithms/meta/Search"
    "muse/util/SequenceBuilder"
    "muse/util/TestRunner"
)

func testBinarySearch() bool {
    size := 32768

    arr := make([]int, size)
    SequenceBuilder.PackIncreasing(arr)

    if Search.BinarySearch(arr, -1) != size {
        return false
    }

    if Search.BinarySearch(arr, 2147483647) != size {
        return false
    }

    for i, v := range arr {
        if Search.BinarySearch(arr, v) != i {
            return false
        }
    }

    return true
}

func testLinearSearch() bool {
    size := 32768

    arr := make([]int, size)
    SequenceBuilder.PackIncreasing(arr)

    if Search.LinearSearch(arr, -1) != size {
        return false
    }

    if Search.LinearSearch(arr, 2147483647) != size {
        return false
    }

    for i, v := range arr {
        if Search.LinearSearch(arr, v) != i {
            return false
        }
    }

    return true
}

func main() {
    TestRunner.ParseTest(testBinarySearch())

    TestRunner.ParseTest(testLinearSearch())
}
